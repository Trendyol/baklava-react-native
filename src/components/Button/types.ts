export type ButtonVariantVariantTypes = 'primary' | 'secondary' | 'tertiary';

export type ButtonKindVariantTypes =
  | 'default'
  | 'neutral'
  | 'success'
  | 'danger';

export type ButtonSizeVariantTypes = 's' | 'm' | 'l';
