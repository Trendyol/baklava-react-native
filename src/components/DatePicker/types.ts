export type Day = {
  day: number;
  month: number;
  year: number;
  isActive: boolean;
  isPrevMonth: boolean;
  isNextMonth: boolean;
  isCurrentDate: boolean;
  timestamp: number;
  yearWithMonth: number;
};

export interface CalendarData {
  calendar: (string[] | Day[])[];
  year: number;
  month: number;
  nameOfMonth: string;
}

export type CalendarOptions = {
  year?: number;
  month?: number;
  nameOfWeekdays?: string[];
  nameOfMonths?: string[];
  firstDayOfWeek?: 0 | 1;
};
