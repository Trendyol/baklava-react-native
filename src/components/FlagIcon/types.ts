import { flagIconList } from './list';

export type FlagIconNameType = (typeof flagIconList)[number];
