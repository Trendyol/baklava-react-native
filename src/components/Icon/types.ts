import { iconList } from './list';

export type IconNameType = typeof iconList[number];
