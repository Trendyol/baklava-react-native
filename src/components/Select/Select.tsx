import React, {
  forwardRef,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import {
  Easing,
  Pressable,
  TouchableOpacity,
  StyleSheet,
  GestureResponderEvent,
} from 'react-native';
import theme, { Theme } from '../../theme';
import Box, { BoxProps } from '../Box/Box';
import Icon from '../Icon/Icon';
import Text from '../Text/Text';
import SelectBottomSheet from '../SelectBottomSheet/SelectBottomSheet';
import { InputLabel } from '../Input/InputLabel';
import { InputHelpText } from '../Input/InputHelpText';
import { useOutlineLabelVisibility } from '../Input/hooks';
import { getBorderColor } from '../Input/utils';
import { VariantProps } from '@ergenekonyigit/restyle';
import { FlagIconNameType } from '../FlagIcon/types';

export interface SelectOption {
  id: string;
  label: string;
  disabled?: boolean;
  flagIcon?: FlagIconNameType | null;
}

export type SelectHandles = {
  focus: () => void;
  blur: () => void;
  clear: () => void;
};

type SelectProps = React.ComponentProps<typeof Box> & {
  label?: string | null;
  subLabel?: string | null;
  labelFixed?: boolean;
  placeholder?: string;
  helpText?: string | null;
  error?: boolean;
  errorText?: string | null;
  success?: boolean;
  successText?: string | null;
  required?: boolean;
  disabled?: boolean;
  size?: VariantProps<Theme, 'inputSizeVariants'>['variant'];
  options: SelectOption[];
  selectedOptions?: string[] | string;
  testID?: string;
  onSelect?: (value: string[]) => void;
  bottomSheetTitle?: string;
  selectButtonLabel?: string;
  closeButtonLabel?: string;
  multiple?: boolean;
  maxVisibleItems?: number;
  bottomSheetPAddingBottom?: BoxProps['pb'];
};

const useSelectState = (selectedOptions: string[] | string) => {
  const [selectedValues, setSelectedValues] = useState<string[]>(
    Array.isArray(selectedOptions) ? selectedOptions : [selectedOptions],
  );

  const bottomSheetSelectionRef = useRef<string[]>([]);

  const resetSelections = useCallback(() => {
    setSelectedValues([]);
  }, []);

  const updateSelections = useCallback((newSelections: string[]) => {
    setSelectedValues(newSelections);
  }, []);

  return {
    selectedValues,
    updateSelections,
    resetSelections,
    bottomSheetSelectionRef,
  };
};

const useDisplayValue = (props: {
  multiple: boolean;
  selectedOptionObjects: SelectOption[];
  selectedValues: string[];
  maxDisplayItems?: number;
}) => {
  const {
    multiple,
    selectedOptionObjects,
    selectedValues,
    maxDisplayItems = 2,
  } = props;

  const getDisplayValues = useCallback(() => {
    if (selectedValues.length === 0) {
      return { visibleText: '', remainingCount: 0 };
    }

    if (multiple) {
      const selectedLabels = selectedOptionObjects.map(option => option.label);

      if (selectedLabels.length <= maxDisplayItems) {
        return {
          visibleText: selectedLabels.join(', '),
          remainingCount: 0,
        };
      } else {
        const visibleLabels = selectedLabels.slice(0, maxDisplayItems);
        const remainingCount = selectedLabels.length - maxDisplayItems;
        return {
          visibleText: visibleLabels.join(', '),
          remainingCount,
        };
      }
    }

    return {
      visibleText:
        selectedOptionObjects.length > 0 ? selectedOptionObjects[0].label : '',
      remainingCount: 0,
    };
  }, [selectedOptionObjects, multiple, selectedValues, maxDisplayItems]);

  return {
    getDisplayValues,
  };
};

const useSortedOptions = (props: {
  options: SelectOption[];
  selectedValues: string[];
  multiple: boolean;
  isBottomSheetVisible: boolean;
  wasCleared: boolean;
}) => {
  const { options, selectedValues, isBottomSheetVisible, wasCleared } = props;
  const originalOptionsRef = useRef<SelectOption[]>([]);

  useEffect(() => {
    if (originalOptionsRef.current.length === 0) {
      originalOptionsRef.current = [...options];
    }
  }, [options]);

  return useMemo(() => {
    if (wasCleared || !isBottomSheetVisible) {
      return options;
    }

    if (selectedValues.length === 0) {
      return options;
    }

    const selectedOptions = options.filter(option =>
      selectedValues.includes(option.id),
    );
    const unselectedOptions = options.filter(
      option => !selectedValues.includes(option.id),
    );

    return [...selectedOptions, ...unselectedOptions];
  }, [options, selectedValues, isBottomSheetVisible, wasCleared]);
};

const Select = forwardRef<SelectHandles, SelectProps>(
  (
    {
      label,
      subLabel,
      labelFixed = false,
      placeholder = '',
      helpText,
      error = false,
      errorText,
      success = false,
      successText,
      required = false,
      disabled = false,
      size = 'medium',
      options = [],
      selectedOptions = [],
      testID = 'select',
      onSelect,
      bottomSheetTitle,
      closeButtonLabel = '',
      selectButtonLabel,
      multiple = true,
      maxVisibleItems = 6,
      bottomSheetPAddingBottom,
      ...rest
    }: SelectProps,
    ref,
  ) => {
    const [focused, setFocused] = useState<boolean>(false);
    const [errorState, setErrorState] = useState<boolean>(error);
    const [successState, setSuccessState] = useState<boolean>(success);
    const [isBottomSheetVisible, setIsBottomSheetVisible] =
      useState<boolean>(false);
    const [wasCleared, setWasCleared] = useState<boolean>(false);

    const {
      selectedValues,
      updateSelections,
      resetSelections,
      bottomSheetSelectionRef,
    } = useSelectState(selectedOptions);

    const selectedOptionObjects = useMemo(
      () => options.filter(option => selectedValues.includes(option.id)),
      [options, selectedValues],
    );

    const sortedOptions = useSortedOptions({
      options,
      selectedValues,
      multiple,
      isBottomSheetVisible,
      wasCleared,
    });

    const inputHeight = theme.inputSizeVariants[size].height;

    const borderColor = useMemo(
      () =>
        getBorderColor({
          focused,
          successState,
          errorState,
        }),
      [focused, successState, errorState],
    );

    const { getDisplayValues } = useDisplayValue({
      multiple,
      selectedOptionObjects,
      selectedValues,
      maxDisplayItems: 2,
    });

    const { startAnimation, animatedViewProps, animatedTextProps } =
      useOutlineLabelVisibility({
        theme,
        easing: Easing.inOut(Easing.ease),
        inputHeight,
        focused,
        value: getDisplayValues().visibleText || '',
        disabled,
        helpText,
        errorText,
        successText,
        errorState,
        successState,
      });

    useEffect(() => {
      setErrorState(error);
      setSuccessState(success);
    }, [error, success]);

    const openBottomSheet = useCallback(async () => {
      if (disabled) {
        return;
      }

      await startAnimation();
      setFocused(true);

      bottomSheetSelectionRef.current = [...selectedValues];

      setWasCleared(false);
      setIsBottomSheetVisible(true);
    }, [disabled, startAnimation, selectedValues, bottomSheetSelectionRef]);

    const closeBottomSheet = useCallback(() => {
      setIsBottomSheetVisible(false);
      setFocused(false);
    }, []);

    const confirmSelection = useCallback(() => {
      const newSelections = [...bottomSheetSelectionRef.current];
      updateSelections(newSelections);

      if (onSelect) {
        onSelect(newSelections);
      }

      closeBottomSheet();
    }, [onSelect, closeBottomSheet, bottomSheetSelectionRef, updateSelections]);

    const handleClear = useCallback(
      (event?: GestureResponderEvent) => {
        event?.stopPropagation();
        resetSelections();

        if (onSelect) {
          onSelect([]);
        }

        setWasCleared(true);
      },
      [onSelect, resetSelections],
    );

    const handleBottomSheetSelection = useCallback(
      (selectedIds: string[]) => {
        bottomSheetSelectionRef.current = selectedIds;
      },
      [bottomSheetSelectionRef],
    );

    const { visibleText, remainingCount } = useMemo(
      () => getDisplayValues(),
      [getDisplayValues],
    );

    const hasSelectedValue = useMemo(
      () => selectedValues.length > 0,
      [selectedValues],
    );

    const iconColor = useMemo(() => {
      if (disabled) {
        return 'neutralLight';
      } else if (focused || hasSelectedValue) {
        return 'neutralDarker';
      } else {
        return 'neutralLight';
      }
    }, [focused, hasSelectedValue, disabled]);

    React.useImperativeHandle(ref, () => ({
      focus: openBottomSheet,
      blur: closeBottomSheet,
      clear: () => {
        handleClear();
      },
    }));

    return (
      <Box py="2xs" {...rest}>
        <InputLabel
          label={label}
          subLabel={subLabel}
          labelFixed={labelFixed}
          errorState={errorState}
          successState={successState}
          animatedViewProps={animatedViewProps}
          animatedTextProps={animatedTextProps}
          inputHeight={inputHeight}
          required={required}
        />

        <Pressable
          onPress={openBottomSheet}
          disabled={disabled}
          testID={`${testID}-pressable`}>
          <Box
            flexDirection="row"
            alignItems="center"
            borderWidth={1}
            borderRadius={'m'}
            borderColor={borderColor}
            backgroundColor={disabled ? 'neutralLightest' : 'neutralFull'}
            px="m"
            height={inputHeight}
            zIndex="layer_0"
            accessibilityLabel={`${testID}-box`}
            testID={`${testID}-box`}>
            <Box
              flex={1}
              flexDirection="row"
              alignItems="center"
              justifyContent="space-between"
              testID={`${testID}-text-container`}>
              <Text
                variant="body2"
                color={visibleText ? 'neutralDarker' : 'neutralLight'}
                testID={`${testID}-value`}
                numberOfLines={1}
                ellipsizeMode="tail"
                style={{ flex: 1 }}>
                {visibleText || placeholder}
              </Text>

              {remainingCount > 0 && (
                <Text
                  variant="body2"
                  color="neutralDarker"
                  testID={`${testID}-remaining-count`}
                  marginRight="m">
                  +{remainingCount}
                </Text>
              )}
            </Box>

            <Box flexDirection="row" alignItems="center">
              {hasSelectedValue && !disabled && (
                <TouchableOpacity
                  onPress={handleClear}
                  hitSlop={{ top: 10, right: 10, bottom: 10, left: 10 }}
                  testID={`${testID}-clear`}
                  style={styles.clearButton}>
                  <Icon name="close" size="s" color="neutralDarker" />
                </TouchableOpacity>
              )}
              {hasSelectedValue && !disabled && (
                <Box
                  height={16}
                  width={1}
                  backgroundColor="neutralLighter"
                  marginHorizontal="xs"
                  testID={`${testID}-divider`}
                />
              )}
              <Icon
                name={isBottomSheetVisible ? 'arrow-up' : 'arrow-down'}
                size="xs"
                color={iconColor}
                testID={`${testID}-icon`}
                accessibilityLabel={`${testID}-icon`}
              />
            </Box>
          </Box>
        </Pressable>

        {(helpText || errorText || successText) && (
          <InputHelpText
            helpText={helpText}
            errorState={errorState}
            errorText={errorText}
            successState={successState}
            successText={successText}
          />
        )}

        <SelectBottomSheet
          pb={bottomSheetPAddingBottom}
          visible={isBottomSheetVisible}
          title={bottomSheetTitle}
          closeButtonLabel={closeButtonLabel}
          closeButtonAction={closeBottomSheet}
          selectButtonLabel={selectButtonLabel}
          selectButtonAction={confirmSelection}
          options={sortedOptions}
          type={multiple ? 'checkbox' : 'radio'}
          onSelect={handleBottomSheetSelection}
          selectedOption={
            isBottomSheetVisible
              ? bottomSheetSelectionRef.current
              : selectedValues
          }
          testID={`${testID}-bottomSheet`}
          maxVisibleItems={maxVisibleItems}
        />
      </Box>
    );
  },
);

const styles = StyleSheet.create({
  clearButton: {
    padding: 2,
  },
});

export default Select;
