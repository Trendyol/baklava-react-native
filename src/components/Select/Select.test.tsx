import React from 'react';
import { act, fireEvent, render, waitFor } from '../../test-utils';
import theme from '../../theme';
import Select, { SelectHandles } from './Select';
import { FlagIconNameType } from '../FlagIcon/types';

describe('Select', () => {
  const defaultOptions = [
    { id: '1', label: 'Option 1' },
    { id: '2', label: 'Option 2' },
    { id: '3', label: 'Option 3' },
  ];

  const defaultProps = {
    options: defaultOptions,
    testID: 'select',
    label: 'Select Label',
    placeholder: 'Select item',
    bottomSheetTitle: 'Options',
    closeButtonLabel: 'Cancel',
    selectButtonLabel: 'Select',
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('should render select correctly', () => {
    // when
    const { toJSON } = render(<Select {...defaultProps} />);

    // then
    expect(toJSON()).toMatchSnapshot();
  });

  test('should render select with label correctly', () => {
    // when
    const { getByText } = render(<Select {...defaultProps} />);

    // then
    expect(getByText('Select Label')).toBeTruthy();
  });

  test('should render select with placeholder correctly', () => {
    // when
    const { getByText } = render(<Select {...defaultProps} />);

    // then
    expect(getByText('Select item')).toBeTruthy();
  });

  test('should render select with disabled state correctly', () => {
    // when
    const { getByTestId } = render(<Select {...defaultProps} disabled />);
    const selectBox = getByTestId('select-box');

    // then
    expect(selectBox.props.style[0].backgroundColor).toBe(
      theme.colors.neutralLightest,
    );
  });

  test('should render select with error state correctly', () => {
    // when
    const { getByTestId } = render(<Select {...defaultProps} error />);
    const selectBox = getByTestId('select-box');

    // then
    expect(selectBox.props.style[0].borderColor).toBe(theme.colors.dangerKey);
  });

  test('should render select with error state and error text correctly', () => {
    // when
    const { getByText } = render(
      <Select {...defaultProps} error errorText="Error message" />,
    );

    // then
    expect(getByText('Error message')).toBeTruthy();
  });

  test('should render select with success state correctly', () => {
    // when
    const { getByTestId } = render(<Select {...defaultProps} success />);
    const selectBox = getByTestId('select-box');

    // then
    expect(selectBox.props.style[0].borderColor).toBe(theme.colors.successKey);
  });

  test('should render select with success state and success text correctly', () => {
    // when
    const { getByText } = render(
      <Select {...defaultProps} success successText="Success message" />,
    );

    // then
    expect(getByText('Success message')).toBeTruthy();
  });

  test('should render select with help text correctly', () => {
    // when
    const { getByText } = render(
      <Select {...defaultProps} helpText="Help message" />,
    );

    // then
    expect(getByText('Help message')).toBeTruthy();
  });

  test('should render select with fixed label correctly', () => {
    // when
    const { toJSON } = render(<Select {...defaultProps} labelFixed />);

    // then
    expect(toJSON()).toMatchSnapshot();
  });

  test('should render select with medium size correctly', () => {
    // when
    const { getByTestId } = render(<Select {...defaultProps} size="medium" />);
    const selectBox = getByTestId('select-box');

    // then
    expect(selectBox.props.style[0].height).toBe(
      theme.inputSizeVariants.medium.height,
    );
  });

  test('should render select with small size correctly', () => {
    // when
    const { getByTestId } = render(<Select {...defaultProps} size="small" />);
    const selectBox = getByTestId('select-box');

    // then
    expect(selectBox.props.style[0].height).toBe(
      theme.inputSizeVariants.small.height,
    );
  });

  test('should render selected option correctly for single select', () => {
    // given
    const mockOptions = [
      { id: '1', label: 'Option 1' },
      { id: '2', label: 'Option 2' },
      { id: '3', label: 'Option 3' },
    ];

    // when
    const { getByText } = render(
      <Select
        testID="select"
        multiple={false}
        selectedOptions="2"
        options={mockOptions}
        placeholder="Select"
      />,
    );

    // then
    expect(getByText('Option 2')).toBeTruthy();
  });

  test('should render selected options correctly for multiple select', () => {
    // when
    const { getByTestId } = render(
      <Select {...defaultProps} multiple selectedOptions={['1', '3']} />,
    );
    const selectValue = getByTestId('select-value');

    // then
    expect(selectValue.props.children).toBe('Option 1, Option 3');
  });

  test('should render +n indicator when selected options exceed maxVisibleItems', () => {
    // when
    const { getByText } = render(
      <Select
        {...defaultProps}
        multiple
        selectedOptions={['1', '2', '3']}
        maxVisibleItems={2}
      />,
    );

    // then
    expect(getByText('+1')).toBeTruthy();
  });

  test('should open bottom sheet when pressed', () => {
    // when
    const { getByTestId } = render(<Select {...defaultProps} />);
    const selectPressable = getByTestId('select-pressable');

    // then
    expect(selectPressable).toBeTruthy();

    fireEvent.press(selectPressable);
    expect(true).toBe(true);
  });

  test('should not open bottom sheet when pressed and disabled', async () => {
    // when
    const { getByTestId, queryByTestId } = render(
      <Select {...defaultProps} disabled />,
    );
    const selectPressable = getByTestId('select-pressable');

    // then
    expect(queryByTestId('select-bottomSheet')).toBeFalsy();

    // when
    fireEvent.press(selectPressable);

    // then
    expect(queryByTestId('select-bottomSheet')).toBeFalsy();
  });

  test('should call onSelect with selected value when item selected in bottom sheet', async () => {
    // given
    jest.useFakeTimers();
    const onSelect = jest.fn();

    // when
    const { getByTestId, getByText } = render(
      <Select {...defaultProps} multiple={false} onSelect={onSelect} />,
    );

    fireEvent.press(getByTestId('select-pressable'));
    await act(async () => {
      jest.runAllTimers();
    });

    // then
    const option = await waitFor(() => getByText('Option 2'));
    fireEvent.press(option);

    fireEvent.press(getByText('Select'));

    expect(onSelect).toHaveBeenCalledWith(['2']);
  });

  test('should call onSelect with selected values when multiple items selected in bottom sheet', async () => {
    // given
    jest.useFakeTimers();
    const onSelect = jest.fn();

    // when
    const { getByTestId, getByText } = render(
      <Select {...defaultProps} multiple onSelect={onSelect} />,
    );

    // Open the bottom sheet
    fireEvent.press(getByTestId('select-pressable'));
    await act(async () => {
      jest.runAllTimers();
    });

    // Selecting individual options
    const option1 = getByText('Option 1');
    fireEvent.press(option1);

    const option3 = getByText('Option 3');
    fireEvent.press(option3);

    // Press the select button
    const selectButton = getByText('Select');
    fireEvent.press(selectButton);
    jest.runAllTimers();

    // then
    expect(onSelect).toHaveBeenCalledWith(['1', '3']);
    jest.useRealTimers();
  });

  test('should clear selected values when clear button pressed', async () => {
    // given
    const onSelect = jest.fn();

    // when
    const { getByTestId } = render(
      <Select
        {...defaultProps}
        multiple
        selectedOptions={['1', '2']}
        onSelect={onSelect}
      />,
    );
    const clearButton = getByTestId('select-clear');
    fireEvent.press(clearButton);

    // then
    expect(onSelect).toHaveBeenCalledWith([]);
  });

  test('should close bottom sheet when close button pressed', async () => {
    // given
    jest.useFakeTimers();

    // when
    const { getByTestId, queryByTestId, getByText } = render(
      <Select {...defaultProps} />,
    );

    // Open the bottom sheet
    fireEvent.press(getByTestId('select-pressable'));
    await act(async () => {
      jest.runAllTimers();
    });

    // Press the close button
    const closeButton = getByText('Cancel');
    fireEvent.press(closeButton);
    jest.runAllTimers();

    // then
    expect(queryByTestId('select-bottomSheet')).toBeFalsy();

    jest.useRealTimers();
  });

  test('should show down arrow when bottom sheet is closed', () => {
    // when
    const { getByTestId } = render(<Select {...defaultProps} />);
    const icon = getByTestId('select-icon');

    // then
    expect(icon.props.title).toBe('arrow-down');
  });

  test('should show up arrow when bottom sheet is open', async () => {
    // when
    const { getByTestId } = render(<Select {...defaultProps} />);
    const selectPressable = getByTestId('select-pressable');
    fireEvent.press(selectPressable);

    // then
    await waitFor(() => {
      const icon = getByTestId('select-icon');
      expect(icon.props.title).toBe('arrow-up');
    });
  });

  test('should render divider when there are selected options', () => {
    // when
    const { getByTestId } = render(
      <Select {...defaultProps} selectedOptions={['1']} />,
    );

    // then
    expect(getByTestId('select-divider')).toBeTruthy();
  });

  test('should not render clear button and divider when disabled', () => {
    // when
    const { queryByTestId } = render(
      <Select {...defaultProps} selectedOptions={['1']} disabled />,
    );

    // then
    expect(queryByTestId('select-clear')).toBeFalsy();
    expect(queryByTestId('select-divider')).toBeFalsy();
  });

  test('should render options with flag icons', async () => {
    // given
    const optionsWithFlags = [
      { id: '1', label: 'Turkey', flagIcon: 'TR' as FlagIconNameType },
      { id: '2', label: 'USA', flagIcon: 'US' as FlagIconNameType },
    ];

    // when
    const { getByTestId } = render(
      <Select
        {...defaultProps}
        options={optionsWithFlags}
        selectedOptions={['1']}
      />,
    );
    const selectPressable = getByTestId('select-pressable');
    fireEvent.press(selectPressable);
  });

  test('should access select handle functions correctly', () => {
    // given
    const ref = React.createRef<SelectHandles>();
    const onSelect = jest.fn();

    // when
    render(<Select {...defaultProps} ref={ref} onSelect={onSelect} />);

    // then
    expect(ref.current?.focus).toBeTruthy();
    expect(ref.current?.blur).toBeTruthy();
    expect(ref.current?.clear).toBeTruthy();

    ref.current?.clear();

    // then
    expect(onSelect).toHaveBeenCalledWith([]);
  });

  test('should not focus or show bottom sheet when openBottomSheet is called through ref on disabled select', () => {
    // given
    jest.useFakeTimers();
    const ref = React.createRef<SelectHandles>();
    const onSelect = jest.fn();

    // when
    const { queryByTestId } = render(
      <Select {...defaultProps} ref={ref} onSelect={onSelect} disabled />,
    );

    // then
    expect(queryByTestId('select-bottomSheet')).toBeFalsy();

    // when
    ref.current?.focus();
    jest.runAllTimers();

    // then
    expect(queryByTestId('select-bottomSheet')).toBeFalsy();
    jest.useRealTimers();
  });

  test('should reset selections when clear method is called', () => {
    // given
    const ref = React.createRef<SelectHandles>();
    const onSelect = jest.fn();

    // when
    render(<Select {...defaultProps} ref={ref} onSelect={onSelect} />);

    ref.current?.clear();

    // then
    expect(onSelect).toHaveBeenCalledWith([]);
  });

  test('selected items should appear at the top of the list when bottom sheet is reopened after selection', async () => {
    // given
    jest.useFakeTimers();
    const onSelect = jest.fn();
    const options = [
      { id: '1', label: 'Option 1' },
      { id: '2', label: 'Option 2' },
      { id: '3', label: 'Option 3' },
      { id: '4', label: 'Option 4' },
    ];

    // when
    const { getByTestId, getByText, queryByTestId, queryAllByTestId } = render(
      <Select
        {...defaultProps}
        options={options}
        multiple
        onSelect={onSelect}
      />,
    );

    fireEvent.press(getByTestId('select-pressable'));
    await act(async () => {
      jest.runAllTimers();
    });

    fireEvent.press(getByText('Option 2'));
    fireEvent.press(getByText('Option 4'));

    fireEvent.press(getByText('Select'));
    jest.runAllTimers();

    expect(queryByTestId('select-bottomSheet')).toBeFalsy();

    fireEvent.press(getByTestId('select-pressable'));
    await act(async () => {
      jest.runAllTimers();
    });

    // then
    const labels = queryAllByTestId('label');
    expect(labels[0].props.children).toContain('Option 2');
    expect(labels[1].props.children).toContain('Option 4');
    expect(labels[2].props.children).toContain('Option 1');
    expect(labels[3].props.children).toContain('Option 3');

    jest.useRealTimers();
  });
});
