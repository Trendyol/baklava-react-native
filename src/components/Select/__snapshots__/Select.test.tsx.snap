// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should render select correctly 1`] = `
<View
  accessible={false}
  style={
    [
      {
        "paddingVertical": 8,
      },
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="none"
    style={
      {
        "bottom": 0,
        "height": 46,
        "left": 12,
        "position": "absolute",
        "zIndex": 2,
      }
    }
  >
    <View
      accessibilityLabel="outlined-label-box"
      accessible={true}
      style={
        [
          {
            "flexDirection": "row",
            "height": 46,
          },
        ]
      }
      testID="outlined-label-box"
    >
      <Text
        collapsable={false}
        style={
          {
            "backgroundColor": "#FFFFFF",
            "color": "#95A1B5",
            "fontFamily": "Rubik-Regular",
            "fontSize": 14,
            "height": 16,
            "lineHeight": 16,
            "paddingLeft": 4,
            "paddingRight": 4,
            "top": 9,
          }
        }
      >
        Select Label
      </Text>
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="select-pressable"
  >
    <View
      accessibilityLabel="select-box"
      accessible={true}
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderColor": "#AFBBCA",
            "borderRadius": 6,
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 40,
            "paddingHorizontal": 16,
            "zIndex": 0,
          },
        ]
      }
      testID="select-box"
    >
      <View
        accessibilityLabel="select-text-container"
        accessible={true}
        style={
          [
            {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
            },
          ]
        }
        testID="select-text-container"
      >
        <Text
          accessibilityLabel="select-value"
          accessible={true}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            [
              {
                "color": "#95A1B5",
                "fontFamily": "Rubik-Regular",
                "fontSize": 14,
                "fontWeight": "400",
                "lineHeight": 16,
                "textAlign": "left",
              },
              {
                "flex": 1,
              },
            ]
          }
          testID="select-value"
        >
          Select item
        </Text>
      </View>
      <View
        accessible={false}
        style={
          [
            {
              "alignItems": "center",
              "flexDirection": "row",
            },
          ]
        }
      >
        <RNSVGSvgView
          accessibilityLabel="select-icon"
          align="xMidYMid"
          bbHeight={14}
          bbWidth={14}
          fill="#95A1B5"
          focusable={false}
          height={14}
          meetOrSlice={0}
          minX={0}
          minY={0}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              {
                "flex": 0,
                "height": 14,
                "width": 14,
              },
            ]
          }
          testID="select-icon"
          title="arrow-down"
          vbHeight={24}
          vbWidth={24}
          width={14}
          xmlns="http://www.w3.org/2000/svg"
        >
          <RNSVGGroup
            fill={
              {
                "payload": 4287996341,
                "type": 0,
              }
            }
            propList={
              [
                "fill",
              ]
            }
          >
            <RNSVGPath
              clipRule={0}
              d="M3.93 7.36c.67-.67 1.756-.67 2.425 0L12 13.003l5.645-5.645a1.714 1.714 0 0 1 2.425 2.425l-6.858 6.857c-.67.67-1.755.67-2.424 0L3.931 9.784a1.714 1.714 0 0 1 0-2.425Z"
              fill={
                {
                  "payload": 4287996341,
                  "type": 0,
                }
              }
              fillRule={0}
              propList={
                [
                  "fill",
                  "fillRule",
                ]
              }
            />
          </RNSVGGroup>
        </RNSVGSvgView>
      </View>
    </View>
  </View>
</View>
`;

exports[`Select should render select with fixed label correctly 1`] = `
<View
  accessible={false}
  style={
    [
      {
        "paddingVertical": 8,
      },
    ]
  }
>
  <View
    accessibilityLabel="fixed-label-box"
    accessible={true}
    style={
      [
        {
          "flexDirection": "row",
          "marginBottom": 4,
        },
      ]
    }
    testID="fixed-label-box"
  >
    <Text
      accessibilityLabel="fixed-label"
      accessible={true}
      style={
        [
          {
            "color": "#6E7787",
            "fontFamily": "Rubik-Medium",
            "fontSize": 12,
            "fontWeight": "500",
            "lineHeight": 14,
            "textAlign": "left",
          },
        ]
      }
      testID="fixed-label"
    >
      Select Label
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="select-pressable"
  >
    <View
      accessibilityLabel="select-box"
      accessible={true}
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderColor": "#AFBBCA",
            "borderRadius": 6,
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 40,
            "paddingHorizontal": 16,
            "zIndex": 0,
          },
        ]
      }
      testID="select-box"
    >
      <View
        accessibilityLabel="select-text-container"
        accessible={true}
        style={
          [
            {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
            },
          ]
        }
        testID="select-text-container"
      >
        <Text
          accessibilityLabel="select-value"
          accessible={true}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            [
              {
                "color": "#95A1B5",
                "fontFamily": "Rubik-Regular",
                "fontSize": 14,
                "fontWeight": "400",
                "lineHeight": 16,
                "textAlign": "left",
              },
              {
                "flex": 1,
              },
            ]
          }
          testID="select-value"
        >
          Select item
        </Text>
      </View>
      <View
        accessible={false}
        style={
          [
            {
              "alignItems": "center",
              "flexDirection": "row",
            },
          ]
        }
      >
        <RNSVGSvgView
          accessibilityLabel="select-icon"
          align="xMidYMid"
          bbHeight={14}
          bbWidth={14}
          fill="#95A1B5"
          focusable={false}
          height={14}
          meetOrSlice={0}
          minX={0}
          minY={0}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              {
                "flex": 0,
                "height": 14,
                "width": 14,
              },
            ]
          }
          testID="select-icon"
          title="arrow-down"
          vbHeight={24}
          vbWidth={24}
          width={14}
          xmlns="http://www.w3.org/2000/svg"
        >
          <RNSVGGroup
            fill={
              {
                "payload": 4287996341,
                "type": 0,
              }
            }
            propList={
              [
                "fill",
              ]
            }
          >
            <RNSVGPath
              clipRule={0}
              d="M3.93 7.36c.67-.67 1.756-.67 2.425 0L12 13.003l5.645-5.645a1.714 1.714 0 0 1 2.425 2.425l-6.858 6.857c-.67.67-1.755.67-2.424 0L3.931 9.784a1.714 1.714 0 0 1 0-2.425Z"
              fill={
                {
                  "payload": 4287996341,
                  "type": 0,
                }
              }
              fillRule={0}
              propList={
                [
                  "fill",
                  "fillRule",
                ]
              }
            />
          </RNSVGGroup>
        </RNSVGSvgView>
      </View>
    </View>
  </View>
</View>
`;
